File: spring-web-java/src/main/java/spring/web/java/application/service/ItemService.java
package spring.web.java.application.service;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.util.Pair;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import lombok.RequiredArgsConstructor;
import spring.web.java.domain.model.entity.Item;
import spring.web.java.domain.repository.ItemRepository;
import spring.web.java.presentation.dto.request.ItemSaveRequest;
import spring.web.java.presentation.dto.response.ItemResponse;
import spring.web.java.presentation.exception.EntityNotFoundException;

@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
public class ItemService {

	private final ItemRepository itemRepository;

	@Transactional
	public ItemResponse saveItem(ItemSaveRequest itemSaveRequest) {
		Item item = itemRepository.save(
			Item.create(
				itemSaveRequest.name(),
				itemSaveRequest.description(),
				itemSaveRequest.price(),
				itemSaveRequest.quantity(),
				itemSaveRequest.category()
			)
		);

		return new ItemResponse(item);
	}

	public Page<ItemResponse> findItems(Pageable pageable, String name) {
		return itemRepository.findAll(pageable, name).map(ItemResponse::new);
	}

	public ItemResponse findItem(Long id) {
		Item item = itemRepository.findById(id)
			.orElseThrow(() -> new EntityNotFoundException(Item.class, id));

		return new ItemResponse(item);
	}

	@Transactional
	public Pair<Boolean, ItemResponse> putItem(Long id, ItemSaveRequest itemSaveRequest) {
		return itemRepository.findById(id)
			.map(item -> Pair.of(false, updateItem(item, itemSaveRequest)))
			.orElseGet(() -> Pair.of(true, saveItem(itemSaveRequest)));
	}

	private ItemResponse updateItem(Item item, ItemSaveRequest itemSaveRequest) {
		item.update(
			itemSaveRequest.name(),
			itemSaveRequest.description(),
			itemSaveRequest.price(),
			itemSaveRequest.quantity(),
			itemSaveRequest.category()
		);

		return new ItemResponse(item);
	}

	@Transactional
	public void deleteItem(Long id) {
		Item item = itemRepository.findById(id)
			.orElseThrow(() -> new EntityNotFoundException(Item.class, id));

		itemRepository.delete(item);
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/application/service/ItemService.kt
package spring.web.kotlin.application.service

import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import spring.web.kotlin.domain.model.entity.Item
import spring.web.kotlin.domain.repository.ItemRepository
import spring.web.kotlin.presentation.dto.request.ItemSaveRequest
import spring.web.kotlin.presentation.dto.response.ItemResponse
import spring.web.kotlin.presentation.exception.EntityNotFoundException

@Service
@Transactional(readOnly = true)
class ItemService(
    private val itemRepository: ItemRepository,
) {
    @Transactional
    fun saveItem(itemSaveRequest: ItemSaveRequest): ItemResponse {
        val item = Item.create(
            itemSaveRequest.name,
            itemSaveRequest.description,
            itemSaveRequest.price,
            itemSaveRequest.quantity,
            itemSaveRequest.category,
        )

        itemRepository.save(item)

        return ItemResponse(item)
    }

    fun findItems(pageable: Pageable, name: String?) =
        itemRepository.findAll(pageable, name).map { ItemResponse(it) }

    fun findItem(id: Long): ItemResponse {
        val item = itemRepository.findByIdOrNull(id) ?: throw EntityNotFoundException(Item::class.java, id)

        return ItemResponse(item)
    }

    @Transactional
    fun putItem(id: Long, itemSaveRequest: ItemSaveRequest) =
        itemRepository.findByIdOrNull(id)
            ?.let { false to updateItem(it, itemSaveRequest) }
            ?: (true to saveItem(itemSaveRequest))

    private fun updateItem(item: Item, itemSaveRequest: ItemSaveRequest): ItemResponse {
        item.update(
            itemSaveRequest.name,
            itemSaveRequest.description,
            itemSaveRequest.price,
            itemSaveRequest.quantity,
            itemSaveRequest.category,
        )

        return ItemResponse(item)
    }

    @Transactional
    fun deleteItem(id: Long) {
        val item = itemRepository.findByIdOrNull(id) ?: throw EntityNotFoundException(Item::class.java, id)

        itemRepository.delete(item)
    }
}
